README.md
# sis8300
device driver for STRUCK SIS8300 board builded on top of upciedev


debian/changelog
sis8300-dkms (1.0.62) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Tue, 08 Nov 2016 09:59:26 +0100

sis8300-dkms (1.0.61) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Tue, 08 Nov 2016 09:31:38 +0100

sis8300-dkms (1.0.60) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Tue, 08 Nov 2016 08:51:38 +0100

sis8300-dkms (1.0.59) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Tue, 08 Nov 2016 08:42:03 +0100

sis8300-dkms (1.0.58) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Mon, 27 Jun 2016 14:19:26 +0200

sis8300-dkms (1.0.57) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Mon, 27 Jun 2016 14:18:40 +0200

sis8300-dkms (1.0.56) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Wed, 15 Jun 2016 09:59:16 +0200

sis8300-dkms (1.0.55) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Fri, 18 Dec 2015 10:13:11 +0100

sis8300-dkms (1.0.54) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Wed, 06 May 2015 10:04:46 +0200

sis8300-dkms (1.0.53) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Wed, 06 May 2015 09:21:09 +0200

sis8300-dkms (1.0.52) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Mon, 16 Mar 2015 10:17:20 +0100

sis8300-dkms (1.0.51) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Mon, 12 Jan 2015 10:48:27 +0100

sis8300-dkms (1.0.50) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Thu, 06 Nov 2014 16:34:14 +0100

sis8300-dkms (1.0.49) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Tue, 21 Oct 2014 10:15:50 +0200

sis8300-dkms (1.0.48) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Tue, 21 Oct 2014 10:14:46 +0200

sis8300-dkms (1.0.47) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Tue, 07 Oct 2014 13:45:11 +0200

sis8300-dkms (1.0.46) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Mon, 03 Mar 2014 08:34:30 +0100

sis8300-dkms (1.0.45) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Tue, 18 Feb 2014 15:42:47 +0100

sis8300-dkms (1.0.44) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Tue, 18 Feb 2014 15:39:21 +0100

sis8300-dkms (1.0.43) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Mon, 04 Nov 2013 09:50:40 +0100

sis8300-dkms (1.0.42) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Wed, 30 Oct 2013 14:39:03 +0100

sis8300-dkms (1.0.41) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Mon, 07 Oct 2013 11:24:27 +0200

sis8300-dkms (1.0.40) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Tue, 27 Aug 2013 14:51:02 +0200

sis8300-dkms (1.0.39) unstable; urgency=low

  * add BUILD_DEPENDS="pciedev" to sis8300-dkms.dkms
    http://comments.gmane.org/gmane.linux.kernel.dkms.devel/876
    https://github.com/dajhorn/dkms

 -- Gerhard Schleßelmann <gerhard.schlesselmann@desy.de>  Mon, 26 Aug 2013 14:08:53 +0200

sis8300-dkms (1.0.38) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Wed, 29 May 2013 10:12:34 +0200

sis8300-dkms (1.0.37) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Tue, 28 May 2013 08:53:35 +0200

sis8300-dkms (1.0.36) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Fri, 03 May 2013 11:18:53 +0200

sis8300-dkms (1.0.35) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Fri, 03 May 2013 11:10:22 +0200

sis8300-dkms (1.0.34) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Fri, 03 May 2013 11:05:45 +0200

sis8300-dkms (1.0.33) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Fri, 03 May 2013 11:05:14 +0200

sis8300-dkms (1.0.32) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Fri, 03 May 2013 11:04:37 +0200

sis8300-dkms (1.0.31) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Fri, 03 May 2013 10:59:10 +0200

sis8300-dkms (1.0.30) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Fri, 03 May 2013 10:50:13 +0200

sis8300-dkms (1.0.29) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Thu, 25 Apr 2013 11:19:48 +0200

sis8300-dkms (1.0.28) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Thu, 25 Apr 2013 11:13:06 +0200

sis8300-dkms (1.0.27) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Thu, 25 Apr 2013 11:12:03 +0200

sis8300-dkms (1.0.26) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Thu, 25 Apr 2013 11:10:42 +0200

sis8300-dkms (1.0.25) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Thu, 25 Apr 2013 11:08:08 +0200

sis8300-dkms (1.0.24) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Thu, 25 Apr 2013 11:06:02 +0200

sis8300-dkms (1.0.23) unstable; urgency=low

  * new version 

 -- Ludwig Petrosyan <lyudvig.petrosyan@desy.de>  Mon, 11 Feb 2013 18:14:57 +0100


sis8300_drv.c
/**
*Copyright 2016-  DESY (Deutsches Elektronen-Synchrotron, www.desy.de)
*
*This file is part of SIS8300 driver.
*
*SIS8300 is free software: you can redistribute it and/or modify
*it under the terms of the GNU General Public License as published by
*the Free Software Foundation, either version 3 of the License, or
*(at your option) any later version.
*
*SIS8300 is distributed in the hope that it will be useful,
*but WITHOUT ANY WARRANTY; without even the implied warranty of
*MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*GNU General Public License for more details.
*
*You should have received a copy of the GNU General Public License
*along with SIS8300.  If not, see <http://www.gnu.org/licenses/>.
**/

/*
*	Author: Ludwig Petrosyan (Email: ludwig.petrosyan@desy.de)
*/

#include <linux/module.h>
#include <linux/fs.h>	
#include <linux/interrupt.h>
#include <linux/sched.h>
#include <linux/types.h>
#include <linux/timer.h>
#include <linux/signal.h>

#include "sis8300_fnc.h"
#include "sis8300_defs.h"
#include "sis8300_reg.h"

MODULE_AUTHOR("Lyudvig Petrosyan");
MODULE_DESCRIPTION("SIS8300 board driver");
MODULE_VERSION("7.1.0");
MODULE_LICENSE("Dual BSD/GPL");

pciedev_cdev     *sis8300_cdev_m = 0;
sis8300_dev       *sis8300_dev_p[PCIEDEV_NR_DEVS];
sis8300_dev       *sis8300_dev_pp;

static int        sis8300_open( struct inode *inode, struct file *filp );
static int        sis8300_release(struct inode *inode, struct file *filp);
static ssize_t sis8300_read(struct file *filp, char __user *buf, size_t count, loff_t *f_pos);
static ssize_t sis8300_write(struct file *filp, const char __user *buf, size_t count, loff_t *f_pos);
static long     sis8300_ioctl(struct file *filp, unsigned int cmd, unsigned long arg);
static int        sis8300_remap_mmap(struct file *filp, struct vm_area_struct *vma);

struct file_operations sis8300_fops = {
    .owner                   =  THIS_MODULE,
    .read                     =  sis8300_read,
    .write                    =  sis8300_write,
    .unlocked_ioctl    =  sis8300_ioctl,
    .open                    =  sis8300_open,
    .release                =  sis8300_release,
    .mmap                 = sis8300_remap_mmap,
};

static struct pci_device_id sis8300_ids[] = {
    { PCI_DEVICE(SIS8300_VENDOR_ID, SIS8300_DEVICE_ID), },
    { PCI_DEVICE(SIS8300_VENDOR_ID, SIS8300L_DEVICE_ID), },
    { 0, }
};
MODULE_DEVICE_TABLE(pci, sis8300_ids);

/*
 * The top-half interrupt handler.
 */
#if LINUX_VERSION_CODE < 0x20613 // irq_handler_t has changed in 2.6.19
static irqreturn_t sis8300_interrupt(int irq, void *dev_id, struct pt_regs *regs)
#else
static irqreturn_t sis8300_interrupt(int irq, void *dev_id)
#endif
{
    uint32_t intreg = 0;
    void*                   address;
    
    struct pciedev_dev *pdev   = (pciedev_dev*)dev_id;
    struct sis8300_dev *dev     = (sis8300_dev *)(pdev->dev_str);
    
    address = pciedev_get_baraddress(BAR0, pdev);
    
    intreg       = ioread32(address + IRQ_STATUS*4);
	
	//printk(KERN_ALERT "SIS8300-PCIEDEV_GET INT IRQ_STATUS %X \n", intreg);
	
    smp_rmb();
     if(intreg == 0){
        return IRQ_NONE;
    }
    if(dev->waitFlag){
        iowrite32(intreg, (address + IRQ_CLEAR*4));
        return IRQ_HANDLED;
    }
    iowrite32(intreg, (address + IRQ_CLEAR*4));
    if(intreg & 0x3){
		dev->waitFlag = 1;
		wake_up_interruptible(&(dev->waitDMA));
    }
    return IRQ_HANDLED;
}

#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,8,0)
    static int sis8300_probe(struct pci_dev *dev, const struct pci_device_id *id)
#else 
static int __devinit sis8300_probe(struct pci_dev *dev, const struct pci_device_id *id)
#endif
{
    int result                = 0;
    int tmp_brd_num  = -1;
    u32 tmp_info          = 0;
    pciedev_dev       *sis8300_pcie_dev;
    void*                   address;
    
    printk(KERN_ALERT "SIS8300-PCIEDEV_PROBE CALLED \n");
    result = pciedev_probe_exp(dev, id, &sis8300_fops, sis8300_cdev_m, SIS8300DEVNAME, &tmp_brd_num);
    printk(KERN_ALERT "SIS8300-PCIEDEV_PROBE_EXP CALLED  FOR BOARD %i result %i\n", tmp_brd_num, result);
    if(!(sis8300_cdev_m->pciedev_dev_m[tmp_brd_num]->pciedev_all_mems)){
        printk(KERN_ALERT "SIS8300-PCIEDEV_PROBE CALLED; NO BARs \n");
        result = pciedev_remove_exp(dev,  sis8300_cdev_m, SIS8300DEVNAME, &tmp_brd_num);
        printk(KERN_ALERT "SIS8300-PCIEDEV_REMOVE_EXP CALLED  FOR SLOT %i\n", tmp_brd_num);  
        return -ENOMEM;
    }
    /*if board has created we will create our structure and pass it to pcedev_dev*/
    if(!result){
        printk(KERN_ALERT "SIS8300-PCIEDEV_PROBE_EXP CREATING CURRENT STRUCTURE FOR BOARD %i\n", tmp_brd_num);
        sis8300_pcie_dev = sis8300_cdev_m->pciedev_dev_m[tmp_brd_num];
        sis8300_dev_pp = kzalloc(sizeof(sis8300_dev), GFP_KERNEL);
        if(!sis8300_dev_pp){
                return -ENOMEM;
        }
        printk(KERN_ALERT "SIS8300-PCIEDEV_PROBE CALLED; CURRENT STRUCTURE CREATED \n");
        sis8300_dev_p[tmp_brd_num] = sis8300_dev_pp;
        sis8300_dev_pp->brd_num      = tmp_brd_num;
        sis8300_dev_pp->parent_dev  = sis8300_cdev_m->pciedev_dev_m[tmp_brd_num];
        init_waitqueue_head(&sis8300_dev_pp->waitDMA);
        pciedev_set_drvdata(sis8300_cdev_m->pciedev_dev_m[tmp_brd_num], sis8300_dev_p[tmp_brd_num]);
        pciedev_setup_interrupt(sis8300_interrupt, sis8300_cdev_m->pciedev_dev_m[tmp_brd_num], SIS8300DEVNAME, 0); 
        
        /*****Switch ON USER_LED*****/
        //address = sis8300_cdev_m->pciedev_dev_m[tmp_brd_num]->memmory_base0;
        address = pciedev_get_baraddress(BAR0, sis8300_pcie_dev);
        iowrite32(0x1, (address + SIS8300_USER_CONTROL_STATUS_REG*4));
        smp_wmb();
        /*Collect INFO*/
        tmp_info = ioread32(address + 0x0);
        smp_rmb();
        sis8300_cdev_m->pciedev_dev_m[tmp_brd_num]->brd_info_list.PCIEDEV_BOARD_ID = tmp_info;
        
        sis8300_dev_pp->fpga_1_gb =0;
        sis8300_dev_pp->dual_optical_interface = 0;
        sis8300_dev_pp->dual_port_14_15_interface = 0;
        sis8300_dev_pp->dual_port_12_13_interface = 0;
        sis8300_dev_pp->dual_channel_sampling = 0;
        sis8300_dev_pp->ringbuffer_delay = 0;
        sis8300_dev_pp->trigger_block_enable = 0;
        tmp_info = ioread32(address + 0x0);
        smp_rmb();
        sis8300_dev_pp->sis8300_mem_max_size = 536870912;
        sis8300_dev_pp->fpga_1_gb = (tmp_info >> 8) & 0x1;
        sis8300_dev_pp->dual_optical_interface = (tmp_info >> 6) & 0x1;
        sis8300_dev_pp->dual_port_14_15_interface = (tmp_info >> 5) & 0x1;
        sis8300_dev_pp->dual_port_12_13_interface = (tmp_info >> 4) & 0x1;
        sis8300_dev_pp->dual_channel_sampling = (tmp_info >> 2) & 0x1;
        sis8300_dev_pp->ringbuffer_delay = (tmp_info >> 1) & 0x1;
        sis8300_dev_pp->trigger_block_enable = tmp_info & 0x1;
        if(sis8300_dev_pp->fpga_1_gb ) sis8300_dev_pp->sis8300_mem_max_size = 1073741824;
/*
        sis8300_cdev_m->pciedev_dev_m[tmp_brd_num]->brd_info_list.PCIEDEV_BOARD_ID = (tmp_info >> 16) & 0xFFFF;
        sis8300_cdev_m->pciedev_dev_m[tmp_brd_num]->brd_info_list.PCIEDEV_BOARD_VERSION = (tmp_info >> 8)  & 0xFFFF;
*/
        tmp_info = ioread32(address + 0x4);
        smp_rmb();
        sis8300_cdev_m->pciedev_dev_m[tmp_brd_num]->brd_info_list.PCIEDEV_HW_VERSION = tmp_info & 0xFFFF;
     }
    return result;
}

#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,8,0)
static void sis8300_remove(struct pci_dev *dev)
#else
static void __devexit sis8300_remove(struct pci_dev *dev)
#endif
{
    int result               = 0;
    int tmp_slot_num = -1;
    int tmp_brd_num = -1;
    printk(KERN_ALERT "SIS8300-REMOVE CALLED\n");
    tmp_brd_num =pciedev_get_brdnum(dev);
    printk(KERN_ALERT "SIS8300-REMOVE CALLED FOR BOARD %i\n", tmp_brd_num);
    /* clean up any allocated resources and stuff here */
    kfree(sis8300_dev_p[tmp_brd_num]);
    /*now we can call pciedev_remove_exp to clean all standard allocated resources
    will clean all interrupts if it seted 
    */
    result = pciedev_remove_exp(dev,  sis8300_cdev_m, SIS8300DEVNAME, &tmp_slot_num);
    printk(KERN_ALERT "SIS8300-PCIEDEV_REMOVE_EXP CALLED  FOR SLOT %i\n", tmp_slot_num);  
}

/****************************************************************************************/
#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,8,0)
static struct pci_driver pci_sis8300_driver = {
    .name       = SIS8300DEVNAME,
    .id_table   = sis8300_ids,
    .probe      = sis8300_probe,
    .remove   = sis8300_remove,
};
#else
static struct pci_driver pci_sis8300_driver = {
    .name       = SIS8300DEVNAME,
    .id_table   = sis8300_ids,
    .probe      = sis8300_probe,
    .remove   = __devexit_p(sis8300_remove),
};
#endif

static int sis8300_open( struct inode *inode, struct file *filp )
{
    int    result = 0;
    result = pciedev_open_exp( inode, filp );
    return result;
}

static int sis8300_release(struct inode *inode, struct file *filp)
{
     int result            = 0;
     result = pciedev_release_exp(inode, filp);
     return result;
} 

static ssize_t sis8300_read(struct file *filp, char __user *buf, size_t count, loff_t *f_pos)
{
     ssize_t    retval         = 0;
    retval  = pciedev_read_exp(filp, buf, count, f_pos);
    return retval;
}

static ssize_t sis8300_write(struct file *filp, const char __user *buf, size_t count, loff_t *f_pos)
{
    ssize_t         retval = 0;
    retval = pciedev_write_exp(filp, buf, count, f_pos);
    return retval;
}

static int sis8300_remap_mmap(struct file *filp, struct vm_area_struct *vma)
{
	ssize_t         retval = 0;
	//printk(KERN_ALERT "PCIEDEV_MMAP CALLED\n");
	retval =pciedev_remap_mmap_exp(filp, vma);
	//printk(KERN_ALERT "PCIEDEV_MMAP_EXP CALLED\n");
	return 0;
}

static long  sis8300_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
{
    long result = 0;
    
    //printk (KERN_ALERT "SIS8300_READ_DMA: START DMA READ 0\n");
    if (_IOC_TYPE(cmd) == PCIEDOOCS_IOC){
		//printk (KERN_ALERT "SIS8300_READ_DMA: START DMA READ 1\n");
        if (_IOC_NR(cmd) <= PCIEDOOCS_IOC_MAXNR && _IOC_NR(cmd) >= PCIEDOOCS_IOC_MINNR) {
			if (_IOC_NR(cmd) <= PCIEDOOCS_IOC_DMA_MAXNR && _IOC_NR(cmd) >= PCIEDOOCS_IOC_DMA_MINNR) {
				//printk (KERN_ALERT "SIS8300_READ_DMA: START DMA READ 3\n");
                result = sis8300_ioctl_dma(filp, &cmd, &arg);
            }else{
			//printk (KERN_ALERT "SIS8300_READ_DMA: START DMA READ 2\n");
            result = pciedev_ioctl_exp(filp, &cmd, &arg, sis8300_cdev_m);
		  }
        }else{
                if (_IOC_NR(cmd) <= SIS8300_IOC_MAXNR && _IOC_NR(cmd) >= SIS8300_IOC_MINNR) {
					//printk (KERN_ALERT "SIS8300_READ_DMA: START DMA READ 4\n");
                        result = sis8300_ioctl_dma(filp, &cmd, &arg);
                    }else{
						//printk (KERN_ALERT "SIS8300_READ_DMA: START DMA READ WRONG ICTL CMD 0\n");
                        return -ENOTTY;
                    }
            }
    }else{
		//printk (KERN_ALERT "SIS8300_READ_DMA: START DMA READ WRONG ICTL CMD 0\n");
         return -ENOTTY;
    }
    return result;
}

static void __exit sis8300_cleanup_module(void)
{
    printk(KERN_NOTICE "SIS8300_CLEANUP_MODULE: PCI DRIVER UNREGISTERED\n");
    pci_unregister_driver(&pci_sis8300_driver);
    printk(KERN_NOTICE "SIS8300_CLEANUP_MODULE CALLED\n");
    upciedev_cleanup_module_exp(&sis8300_cdev_m);
    
}

static int __init sis8300_init_module(void)
{
    int   result = 0;
    
    printk(KERN_WARNING "SIS8300_INIT_MODULE CALLED\n");
    result = upciedev_init_module_exp(&sis8300_cdev_m, &sis8300_fops, SIS8300DEVNAME);
    result = pci_register_driver(&pci_sis8300_driver);
    printk(KERN_ALERT "SIS8300_INIT_MODULE:REGISTERING PCI DRIVER RESUALT %d\n", result);
    return 0; /* succeed */
}

module_init(sis8300_init_module);
module_exit(sis8300_cleanup_module);



sis8300_fnc.h
/**
*Copyright 2016-  DESY (Deutsches Elektronen-Synchrotron, www.desy.de)
*
*This file is part of SIS8300 driver.
*
*SIS8300 is free software: you can redistribute it and/or modify
*it under the terms of the GNU General Public License as published by
*the Free Software Foundation, either version 3 of the License, or
*(at your option) any later version.
*
*SIS8300 is distributed in the hope that it will be useful,
*but WITHOUT ANY WARRANTY; without even the implied warranty of
*MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*GNU General Public License for more details.
*
*You should have received a copy of the GNU General Public License
*along with SIS8300.  If not, see <http://www.gnu.org/licenses/>.
**/

/*
*	Author: Ludwig Petrosyan (Email: ludwig.petrosyan@desy.de)
*/


#ifndef _SIS8300_FNC_H_
#define _SIS8300_FNC_H_
#include <linux/version.h>
#include "pciedev_io.h"
#include "pciedev_ufn.h"

#ifndef SIS8300_NR_DEVS
#define SIS8300_NR_DEVS 15  /* sis83000 through sis830011 */
#endif

#define SIS8300DEVNAME "sis8300"	                    /* name of device */
#define SIS8300_VENDOR_ID 0x1796	/* FZJZEL vendor ID */
#define SIS8300_DEVICE_ID 0x0018	/* SIS8300 dev board device ID */
#define SIS8300L_DEVICE_ID 0x0019	/* SIS8300L dev board device ID */

#define SIS8300_KERNEL_DMA_BLOCK_SIZE 131072    // 128kByte
//#define SIS8300_KERNEL_DMA_BLOCK_SIZE 262144 // 256kByte
//#define SIS8300_MEM_MAX_SIZE          536870912 // 512MByte
//#define SIS8300_DMA_MAX_SYZE         32768
//#define SIS8300_DMA_MIN_SYZE          128
#define SIS8300_DMA_SYZE                  4096

struct sis8300_dev {
    int                              brd_num;
    struct timeval          dma_start_time;
    struct timeval          dma_stop_time;
    int                              waitFlag;
    u32                            dev_dma_size;
    u32                            dma_page_num;
    int                              dma_offset;
    int                              dma_page_order;
    wait_queue_head_t  waitDMA;
    struct pciedev_dev  *parent_dev;
    int                          sis8300_mem_max_size;
    int                          fpga_1_gb;
    int                          dual_optical_interface;
    int                          dual_port_14_15_interface;
    int                          dual_port_12_13_interface;
    int                          dual_channel_sampling;
    int                          ringbuffer_delay;
    int                          trigger_block_enable;
};
typedef struct sis8300_dev sis8300_dev;

long sis8300_ioctl_dma(struct file *, unsigned int* , unsigned long* );

#endif /* _SIS8300_FNC_H_ */


sis8300_defs.h
/* 
 * File:   sis8300_defs.h
 * Author: petros
 *
 * Created on January 12, 2011, 5:20 PM
 */

/**
*Copyright 2016-  DESY (Deutsches Elektronen-Synchrotron, www.desy.de)
*
*This file is part of SIS8300 driver.
*
*SIS8300 is free software: you can redistribute it and/or modify
*it under the terms of the GNU General Public License as published by
*the Free Software Foundation, either version 3 of the License, or
*(at your option) any later version.
*
*SIS8300 is distributed in the hope that it will be useful,
*but WITHOUT ANY WARRANTY; without even the implied warranty of
*MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*GNU General Public License for more details.
*
*You should have received a copy of the GNU General Public License
*along with SIS8300.  If not, see <http://www.gnu.org/licenses/>.
**/

#ifndef SIS8300_DEFS_H
#define	SIS8300_DEFS_H

#include <linux/ioctl.h> /* needed for the _IOW etc stuff used later */
#include "pciedev_io.h"

//#define IOCTRL_R      0x00
//#define IOCTRL_W     0x01
//#define IOCTRL_ALL  0x02
#define MUXA              0x00
#define MUXB              0x01
#define MUXC              0x02
#define MUXD               0x04
#define MUXE               0x05
#define SPI_AD9510     0x41
#define SPI_SI5326       0x42
#define SPI_DAC            0x44
#define SPI_ADC            0x48
#define MODE_RESET  0x4
#define MODE_TRG      0x2
#define MODE_NOW    0x1

struct device_ioctrl_data_buf  {
        u_int    offset;
        u_int    cmd;
        u_int    num;
        u_int    step;
        u_int    data[64];
};
typedef struct device_ioctrl_data_buf device_ioctrl_data_buf;



typedef struct t_sis8300_reg{
	u_int offset; /* offset from bar0 */
	u_int data;   /* data which will be read / written */
}sis8300_reg;

/* Use 'o' as magic number */

#define SIS8300_IOC           			'0'
#define SIS8300_PHYSICAL_SLOT       _IOWR(SIS8300_IOC, 20, int)
#define SIS8300_REG_READ            _IOWR(SIS8300_IOC, 21, int)
#define SIS8300_REG_WRITE           _IOWR(SIS8300_IOC, 22, int)
#define SIS8300_GET_DMA_TIME 	    _IOWR(SIS8300_IOC, 23, int)
#define SIS8300_DRIVER_VERSION      _IOWR(SIS8300_IOC, 24, int)
#define SIS8300_FIRMWARE_VERSION    _IOWR(SIS8300_IOC, 25, int)
#define SIS8300_EXT_TRIGGER         _IOWR(SIS8300_IOC, 26, int)
#define SIS8300_DISABLE_CHANNEL     _IOWR(SIS8300_IOC, 27, int)
#define SIS8300_MUX_SELECT          _IOWR(SIS8300_IOC, 28, int)
#define SIS8300_SPI_INTRFC          _IOWR(SIS8300_IOC, 29, int)
#define SIS8300_DAC_CONTROL         _IOWR(SIS8300_IOC, 30, int)
#define SIS8300_DAC_DATA            _IOWR(SIS8300_IOC, 31, int)
#define SIS8300_TAP_DELAY           _IOWR(SIS8300_IOC, 32, int)
#define SIS8300_TRG_SETUP           _IOWR(SIS8300_IOC, 33, int)
#define SIS8300_THRES_REG           _IOWR(SIS8300_IOC, 34, int)
#define SIS8300_SAMPLE_ADDR         _IOWR(SIS8300_IOC, 35, int)
#define SIS8300_SAMPLE_LENGTH       _IOWR(SIS8300_IOC, 36, int)
#define SIS8300_RINGBUF_DELAY       _IOWR(SIS8300_IOC, 37, int)
#define SIS8300_CNTRL_STS           _IOWR(SIS8300_IOC, 38, int)
#define SIS8300_DMA_OFFSET          _IOWR(SIS8300_IOC, 39, int)
#define SIS8300_READ_DMA            _IOWR(SIS8300_IOC, 40, int)
#define SIS8300_WRITE_DMA           _IOWR(SIS8300_IOC, 41, int)
#define SIS8300_READ_DMA_4k         _IOWR(SIS8300_IOC, 42, int)
#define SIS8300_DDR2_TEST_ENBL      _IOWR(SIS8300_IOC, 43, int)
#define SIS8300_HLINK_IN            _IOWR(SIS8300_IOC, 44, int)
#define SIS8300_HLINK_OUT           _IOWR(SIS8300_IOC, 45, int)
#define SIS8300_HLINK_TRG           _IOWR(SIS8300_IOC, 46, int)
#define SIS8300_HLINK_TRG_FEDGE     _IOWR(SIS8300_IOC, 47, int)
#define SIS8300_MLVDS_OUT           _IOWR(SIS8300_IOC, 48, int)
#define SIS8300_MLVDS_OUT_ENBL      _IOWR(SIS8300_IOC, 49, int)
#define SIS8300_MLVDS_TRG           _IOWR(SIS8300_IOC, 50, int)
#define SIS8300_MLVDS_IN            _IOWR(SIS8300_IOC, 51, int)
#define SIS8300_MLVDS_TRG_EDGE      _IOWR(SIS8300_IOC, 52, int)
#define SIS8300_MLVDS_REG           _IOWR(SIS8300_IOC, 53, int)
#define SIS8300_BLINK_LED           _IOWR(SIS8300_IOC, 54, int)
#define SIS8300_SAMPLE_ADDR_ALL         _IOWR(SIS8300_IOC, 55, int)
#define SIS8300_WRITE_DMA_2PEER         _IOWR(SIS8300_IOC, 56, int)

#define SIS8300_IOC_MAXNR           57
#define SIS8300_IOC_MINNR           20

#endif	/* SIS8300_DEFS_H */

